### PR TITLE
Load ServiceMon secrets directly

### PR DESCRIPTION
**Pull request by redbaron**
Alternative implementation for same problem as in https://github.com/coreos/prometheus-operator/pull/1997

### CODE REVIEW

File: pkg/prometheus/operator.go [hunk 1]
```diff
@@ -1159,28 +1160,60 @@ func loadAdditionalScrapeConfigsSecret(additionalScrapeConfigs *v1.SecretKeySele
 	return nil, nil
 }
 
+func extractCredKey(secret *v1.Secret, sel v1.SecretKeySelector, cred string) (string, error) {
+	if s, ok := secret.Data[sel.Key]; ok {
+		return string(s), nil
+	} else {
```
Comment (squat) [inline pkg/prometheus/operator.go]: nit: there is no need for this `else` here since you are returning early

File: pkg/prometheus/operator.go [hunk 2]
```diff
@@ -1159,28 +1160,60 @@ func loadAdditionalScrapeConfigsSecret(additionalScrapeConfigs *v1.SecretKeySele
...
+	if s, ok = cache[cacheKey]; !ok {
+		if s, err = c.Get(sel.Name, metav1.GetOptions{}); err != nil {
+			return "", fmt.Errorf("unable to fetch %s secret %q: %s", cred, sel.Name, err)
+		}
+		cache[cacheKey] = s
+	}
+	return extractCredKey(s, sel, cred)
+}
+
+func loadBasicAuthSecretFromAPI(basicAuth *monitoringv1.BasicAuth, c corev1client.CoreV1Interface, ns string, cache map[string]*v1.Secret) (_ BasicAuthCredentials, err error) {
```
Comment (squat) [inline pkg/prometheus/operator.go]: nit: we are standardized around not using named returns for this project

File: pkg/prometheus/operator.go [hunk 3]
```diff
@@ -1205,25 +1238,12 @@ func (c *Operator) loadBasicAuthSecrets(
...
-		}
-	}
-
 	secrets := map[string]BasicAuthCredentials{}
+	nsSecretCache := make(map[string]*v1.Secret)
 	for _, mon := range mons {
 		for i, ep := range mon.Spec.Endpoints {
 			if ep.BasicAuth != nil {
-				credentials, err := loadBasicAuthSecret(ep.BasicAuth, sMonSecretMap[mon.Namespace])
+				credentials, err := loadBasicAuthSecretFromAPI(ep.BasicAuth, c.kclient.CoreV1(), mon.Namespace, nsSecretCache)
```
Comment (squat) [inline pkg/prometheus/operator.go]: Rather than unnecessarily GETing secrets from the API when we could be GETing them from the in-memory secrets cache.

Comment (redbaron) [reply pkg/prometheus/operator.go]: This PR is a preparation for https://github.com/coreos/prometheus-operator/pull/2197 , with endgoal of reducing scope of RBAC permissions required for namespaces with ServiceMonitor and PrometheusRule CRs

**Comment by brancz** [comment]
I generally like the approach, I'm a little concerned about the amount of calls this may produce against the Kubernetes API, but as this is bound to the number of secrets specified in servicemonitors I think it's ok for now. The only thing that seems a little strange is that users would have to bind the role to access these secrets to the prometheus operator. How do they know what to bind it to?

File: pkg/prometheus/operator.go (hunk 1 above)
Comment (redbaron) [reply pkg/prometheus/operator.go]: updated

File: pkg/prometheus/operator.go (hunk 2 above)
Comment (redbaron) [reply pkg/prometheus/operator.go]: updated

**Comment by redbaron** [comment]
>  but as this is bound to the number of secrets specified in servicemonitors I think it's ok for now. 

there is a per-iteration caching to to try to minimize it, I doubt it is going to be effective though

> The only thing that seems a little strange is that users would have to bind the role to access these secrets to the prometheus operator service account. How do they know what to bind it to?

Doesn't prometheus-operator update `.status` field? It can add error/success reports there so that users can know if their SMon was processed and what went wrong.

File: pkg/prometheus/operator.go (hunk 3 above)
Comment (squat) [reply pkg/prometheus/operator.go]: Got it :+1:  there was no mention of 2197 so it was not clear why the lister was not used

**Comment by redbaron** [comment]
I think all comments are addressed now, is it a good time to merge it? I have more PRs regarding reducing RBAC scope and working within restricted namespaces and would like to keep number of them in-flight to a bare minimum

**Comment by brancz** [comment]
Sorry for taking so long, KubeCon kept me busy real good. This lgtm :+1:

**Pull request merged**